### ISSUE TITLE
declare global and no-default-lib cause tsc to not emit modules

### ISSUE THREAD

**Issue by satyarohith** [denoland/deno#14093]
Steps to reproduce: 
```
$ git clone --branch upgrade_std git@github.com:satyarohith/sift.git
$ cd sift && deno run tst.ts
```
Panic:
```
Platform: macos aarch64
Version: 1.20.1
Args: ["deno", "run", "tst.ts"]

thread 'main' panicked at 'internal error: entered unreachable code: Unexpected missing emit: file:///Users/sr/sift/mod.ts', cli/proc_state.rs:585:25
stack backtrace:
   0:        0x1035e69ec - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h58c637f9e30d1b8a
   1:        0x102ec3b78 - core::fmt::write::hc16ca97d9f5e0acd
   2:        0x1035e20c0 - std::io::Write::write_fmt::hd53dc7153ccce04d
   3:        0x1035eb6e4 - std::panicking::default_hook::{{closure}}::ha41f2cf175ca93cf
   4:        0x1035eb35c - std::panicking::default_hook::h838498b157804a1a
   5:        0x102d69b14 - deno::setup_panic_hook::{{closure}}::h31e91d21bbeca7aa
   6:        0x1035ec21c - std::panicking::rust_panic_with_hook::h7722806c30430655
   7:        0x1035ebf80 - std::panicking::begin_panic_handler::{{closure}}::h265229fa7cad6de7
   8:        0x1035ebf00 - std::sys_common::backtrace::__rust_end_short_backtrace::hfa83d61d0ed14557
   9:        0x1035ebecc - _rust_begin_unwind
  10:        0x102ec255c - core::panicking::panic_fmt::hd3f68d55261c2af0
  11:        0x102d205c4 - deno::proc_state::ProcState::load::h49296864fdb468de
  12:        0x102da5718 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h2f5e601905fb22c8
  13:        0x102f5bbb0 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hcd6454cf409a029c
  14:        0x102f60224 - <deno_core::modules::RecursiveModuleLoad as futures_core::stream::Stream>::poll_next::hf1912edce6c880f3
  15:        0x102dc38c4 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h8aa0a6ae387fc16e
  16:        0x102daf038 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h48668c9f6d6fd0d7
  17:        0x102dd4048 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hbd8846c1b035b27e
  18:        0x102d60b74 - deno::run_command::{{closure}}::he5da6270f4029a65
  19:        0x102db2304 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h55d3a810425f8cda
  20:        0x102d6c568 - deno::main::{{closure}}::h7612d26431395f50
  21:        0x102dad320 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h3ad9cb18410bcf24
  22:        0x102cabb6c - deno_runtime::tokio_util::run_basic::hdc07c2787f941f2e
  23:        0x102d69c80 - deno::main::h446f202d500b3cb0
  24:        0x102d882cc - std::sys_common::backtrace::__rust_begin_short_backtrace::h6f18f9884c615993
  25:        0x102dfe96c - _main
```

**Comment by satyarohith** [comment]
Alternative reproduction:

`main.ts`
```ts
import {
  serve,
  Status,
  STATUS_TEXT,
} from "https://deno.land/x/sift@0.4.4/mod.ts";

serve({
  "/api": () => new Response("WIP"),
  404: () =>
    new Response(STATUS_TEXT.get(Status.BadRequest), {
      status: Status.BadRequest,
    }),
}, {
  port: 8910,
});

```
Run:
```
$ deno run -A main.ts
```

Output:

```
deno run -A main.ts
Check file:///Users/sr/c/github.com/satyarohith/sift/main.ts

============================================================
Deno has panicked. This is a bug in Deno. Please report this
at https://github.com/denoland/deno/issues/new.
If you can reliably reproduce this panic, include the
reproduction steps and re-run with the RUST_BACKTRACE=1 env
var set and include the backtrace in your report.

Platform: macos aarch64
Version: 1.20.1
Args: ["deno", "run", "-A", "main.ts"]

thread 'main' panicked at 'internal error: entered unreachable code: Unexpected missing emit: https://deno.land/x/sift@0.4.4/mod.ts', cli/proc_state.rs:585:25
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by dsherret** [comment]
@kitsonk and I looked into this one today in our pair programming.

Simplified reproduction:

```ts
/// <reference no-default-lib="true"/>
/// <reference lib="es5" />

declare global {
  interface Test {
  }
}

export {};
```

Essentially, it looks like TypeScript doesn't emit this file because it affects the global scope (`isFileAffectingGlobalScope`) AND it contains a `no-default-lib="true"` (`isSourceFileDefaultLibrary` where `sourceFile.hasNoDefaultLib` is true). Temporary workaround is to move this code into a separate .d.ts file that doesn't need to be emitted.

I wasn't able to reproduce it with the vanilla tsc on the command line though. I will look into it more tomorrow. Opened https://github.com/microsoft/TypeScript/pull/48412 in the meantime.

**Comment by dsherret** [comment]
Opened https://github.com/microsoft/TypeScript/issues/48423

**Comment by nayeemrmn** [comment]
I think this was fixed in #15022

**Comment by dsherret** [comment]
Yup. Thanks @nayeemrmn!
